File: clients/csharp/src/Neo4j.AgentMemory/MemoryClient.cs
using Neo4j.AgentMemory.Auth;
using Neo4j.AgentMemory.LongTerm;
using Neo4j.AgentMemory.Models;
using Neo4j.AgentMemory.Query;
using Neo4j.AgentMemory.Reasoning;
using Neo4j.AgentMemory.ShortTerm;
using Neo4j.AgentMemory.Transport;

namespace Neo4j.AgentMemory;

/// <summary>
/// Client for neo4j-agent-memory. Picks the transport automatically:
///
///   • REST when the endpoint contains a /v1 segment (hosted service).
///   • Bridge otherwise (TCK conformance servers, local reference).
///
/// Override the choice with <see cref="MemoryClientOptions.Transport"/>.
/// </summary>
public class MemoryClient : IAsyncDisposable, IDisposable
{
    private readonly ITransport _transport;
    private readonly bool _ownsTransport;

    public ShortTermMemory ShortTerm { get; }
    public LongTermMemory LongTerm { get; }
    public ReasoningMemory Reasoning { get; }
    public QueryConsole Query { get; }
    public AuthClient Auth { get; }

    public MemoryClient(MemoryClientOptions options)
    {
        _transport = BuildTransport(options);
        _ownsTransport = true;
        ShortTerm = new ShortTermMemory(_transport);
        LongTerm = new LongTermMemory(_transport);
        Reasoning = new ReasoningMemory(_transport);
        Query = new QueryConsole(_transport);
        Auth = new AuthClient(_transport);
    }

    public MemoryClient(ITransport transport)
    {
        _transport = transport;
        _ownsTransport = false;
        ShortTerm = new ShortTermMemory(_transport);
        LongTerm = new LongTermMemory(_transport);
        Reasoning = new ReasoningMemory(_transport);
        Query = new QueryConsole(_transport);
        Auth = new AuthClient(_transport);
    }

    private static ITransport BuildTransport(MemoryClientOptions options)
    {
        var mode = options.Transport;
        if (mode == TransportMode.Auto)
            mode = LooksLikeRest(options.Endpoint) ? TransportMode.Rest : TransportMode.Bridge;

        return mode switch
        {
            TransportMode.Rest => new RestTransport(options.Endpoint, options.ApiKey, options.Timeout, options.TokenProvider, options.Headers),
            _ => new BridgeTransport(options.Endpoint, options.ApiKey, options.Timeout, options.Headers),
        };
    }

    private static bool LooksLikeRest(string endpoint)
    {
        if (string.IsNullOrEmpty(endpoint)) return false;
        foreach (var segment in endpoint.Split('/'))
        {
            if (segment.Length >= 2 && segment[0] == 'v' &&
                int.TryParse(segment.AsSpan(1), out _))
                return true;
        }
        return false;
    }

    public Task ConnectAsync(CancellationToken ct = default) => _transport.ConnectAsync(ct);
    public Task CloseAsync(CancellationToken ct = default) => _transport.CloseAsync(ct);

    public Task ClearAllDataAsync(CancellationToken ct = default) =>
        _transport.RequestAsync("clear_all_data", null, ct);

    public async ValueTask DisposeAsync()
    {
        await CloseAsync();
        if (_ownsTransport && _transport is IDisposable d) d.Dispose();
        GC.SuppressFinalize(this);
    }

    public void Dispose()
    {
        if (_ownsTransport && _transport is IDisposable d) d.Dispose();
        GC.SuppressFinalize(this);
    }
}

File: clients/rlang/neo4j.memory/R/memory_client.R
#' @title MemoryClient — root entry point
#'
#' @description
#' Picks the transport automatically based on the endpoint shape:
#'
#' * REST when the endpoint contains a `/v1` segment (hosted service).
#' * Bridge otherwise (TCK conformance servers, local reference).
#'
#' Override with `transport_mode = "bridge"` or `"rest"`.
#'
#' @export
MemoryClient <- R6::R6Class("MemoryClient",
  public = list(
    short_term = NULL,
    long_term = NULL,
    reasoning = NULL,
    query = NULL,
    auth = NULL,

    #' @param endpoint Service URL (`http://localhost:3001` or
    #'   `https://memory.neo4jlabs.com/v1`).
    #' @param api_key Bearer API key (e.g. `nams_*`).
    #' @param transport Pre-built transport instance — overrides endpoint.
    #' @param transport_mode "auto", "bridge", or "rest".
    #' @param token_provider Optional function returning a fresh token (REST).
    #' @param headers Named list of additional headers.
    #' @param timeout Request timeout in seconds (default 30).
    initialize = function(endpoint = NULL, api_key = NULL, transport = NULL,
                          transport_mode = "auto",
                          token_provider = NULL, headers = NULL,
                          timeout = 30) {
      if (!is.null(transport)) {
        private$transport <- transport
      } else if (!is.null(endpoint)) {
        mode <- transport_mode
        if (mode == "auto") {
          mode <- if (grepl("/v[0-9]+(/|$)", endpoint)) "rest" else "bridge"
        }
        private$transport <- if (mode == "rest") {
          RestTransport$new(endpoint, api_key = api_key, timeout = timeout,
            token_provider = token_provider, headers = headers)
        } else {
          BridgeTransport$new(endpoint, api_key = api_key, timeout = timeout,
            headers = headers)
        }
      } else {
        stop("Either 'endpoint' or 'transport' must be provided", call. = FALSE)
      }

      self$short_term <- ShortTermMemory$new(private$transport)
      self$long_term <- LongTermMemory$new(private$transport)
      self$reasoning <- ReasoningMemory$new(private$transport)
      self$query <- QueryConsole$new(private$transport)
      self$auth <- AuthClient$new(private$transport)
    },

    connect = function() { private$transport$connect(); invisible(self) },
    close = function() { private$transport$close(); invisible(NULL) },

    clear_all_data = function() {
      tryCatch(private$transport$request("clear_all_data"),
        error = function(e) NULL)
      invisible(NULL)
    }
  ),

  private = list(transport = NULL)
)
